### PR TITLE
feat: 🎸 Add Asset Metadata endpoints

### PR DESCRIPTION
**Pull request by prashantasdeveloper**
### JIRA Link 

DA-446, DA-447, DA-448, DA-449, DA-451, DA-452

### Changelog / Description 

Add Asset Metadata endpoints

### Checklist - 

- [x] New Feature ?
- [x] Updated swagger annotation (if API structure is changed) ?
- [x] Unit Test (if possible) ?
- [ ] Updated the Readme.md (if required) ?

### CODE REVIEW

File: src/metadata/metadata.controller.spec.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+      expect(result).toEqual({
+        results: [new MetadataEntryModel({ asset: ticker, type, id })],
+      });
+    });
+  });
+
+  describe('getSingleMetadata', () => {
+    it('should return the Metadata details for a specific type and ID', async () => {
+      const mockMetadataEntry = new MockMetadataEntry();
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.spec.ts:68]: - use `createMock`, specify the assertion values
- remove `as any` cast

`mockResolvedValue` should be able to stay as they are

File: src/metadata/metadata.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      );
+    });
+  });
+
+  describe('createMetadata', () => {
+    it('should accept CreateMetadataDto and create a local Asset Metadata for the given ticker', async () => {
+      const transaction = {
+        blockHash: '0x1',
+        transactionHash: '0x2',
+        blockNumber: new BigNumber(1),
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.spec.ts:110]: (optional) Just an idea, but we could put these into `src/test-utils/consts.ts` and call it something like `txBase` and do something like this:

```suggestion
        ...txBase
```

OK to leave as it though

File: src/metadata/metadata.controller.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      'This endpoint creates a local metadata for the given Asset. The metadata value can be set by passing `value` parameter and specifying other optional `details` about the value',
+  })
+  @ApiParam({
+    name: 'ticker',
+    description: 'The ticker of the Asset for which the metadata is to be created',
+    type: 'string',
+    example: 'TICKER',
+  })
+  @ApiTransactionResponse({
+    description: 'Newly created Metadata entry along with transaction details',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:103]:
```suggestion
    description: 'The newly created Metadata entry along with transaction details',
```

File: src/metadata/metadata.controller.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    type: CreatedMetadataEntryModel,
+  })
+  @ApiTransactionFailedResponse({
+    [HttpStatus.NOT_FOUND]: ['The Asset was not found'],
+    [HttpStatus.BAD_REQUEST]: [
+      'Asset Metadata name length exceeded',
+      'Asset Metadata value length exceeded',
+    ],
+    [HttpStatus.UNPROCESSABLE_ENTITY]: [
+      'Asset Metadata with the given name already exists',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:113]: Name exists should be `409 Conflict` imo

File: src/metadata/metadata.controller.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      new CreatedMetadataEntryModel({
+        transactions,
+        metadata: new MetadataEntryModel({ asset: assetTicker, id, type }),
+      });
+
+    return handleServiceResult(serviceResult, resolver);
+  }
+
+  @ApiOperation({
+    summary: 'Set Asset Metadata value and/or its details',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:142]:
```suggestion
    summary: 'Set an Asset's Metadata value and details',
```

File: src/metadata/metadata.controller.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    description:
+      'This endpoint assigns a new value for the Metadata along with its details or optionally only set the details (expiry + lock status) of the Metadata value. Note that the value of a locked Metadata cannot be altered',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:144]:
```suggestion
      'This endpoint assigns a new value for the Metadata along with its expiry and lock status (when provided with `details`) of the Metadata value. Note that the value of a locked Metadata cannot be altered',
```

File: src/metadata/metadata.controller.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+  })
+  @ApiParam({
+    name: 'ticker',
+    description: 'The ticker of the Asset for which the metadata value is to be set',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:148]:
```suggestion
    description: 'The ticker of the Asset for which the Metadata value is to be set',
```

File: src/metadata/metadata.controller.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    description: 'Details about the transaction',
+    type: TransactionQueueModel,
+  })
+  @ApiTransactionFailedResponse({
+    [HttpStatus.NOT_FOUND]: ['The Asset was not found'],
+    [HttpStatus.BAD_REQUEST]: ['Asset Metadata name length exceeded'],
+    [HttpStatus.UNPROCESSABLE_ENTITY]: [
+      'Details cannot be set for a locked Metadata value',
+      'Metadata value is currently locked',
+      'Metadata value details cannot be set for a metadata with no value',
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:174]:
```suggestion
      'Details cannot be set for a metadata without a value',
```
I think the sentence structure should be consistent with the locked error.

File: src/metadata/metadata.controller.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    description: 'The ID of Asset Metadata',
+    type: 'string',
+    example: '1',
+  })
+  @ApiTransactionResponse({
+    description: 'Details about the transaction',
+    type: TransactionQueueModel,
+  })
+  @ApiTransactionFailedResponse({
+    [HttpStatus.NOT_FOUND]: ['The Asset was not found'],
```
Comment (polymath-eric) [inline src/metadata/metadata.controller.ts:169]: Could the Metadata not be found as well?

File: src/metadata/metadata.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+        let error;
+        try {
+          await service.findOne({ ticker, type, id });
+        } catch (err) {
+          error = err;
+        }
+
+        expect(error).toBeInstanceOf(NotFoundException);
+      });
+    });
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:147]:
```suggestion
    });
    
```

File: src/metadata/metadata.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        mockAsset.metadata.getOne.mockImplementation(() => {
+          throw expectedError;
+        });
+
+        let error;
+        try {
+          await service.findOne({ ticker, type, id });
+        } catch (err) {
+          error = err;
+        }
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:160]:
```suggestion
  await expect(service.findOne({ticker, type, id}).rejects.toThrow(expectedError)
```

File: src/metadata/metadata.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+
+        expect(error).toEqual(expectedError);
+
+        expectedError = new Error('Something else');
+
+        mockIsPolymeshError.mockReturnValue(true);
+
+        error = null;
+        try {
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:169]: `expect().rejects` syntax here to

File: src/metadata/metadata.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+          await service.findOne({ ticker, type, id });
+        } catch (err) {
+          error = err;
+        }
+
+        expect(error).toEqual(expectedError);
+      });
+    });
+    describe('otherwise', () => {
+      it('should return the Asset', async () => {
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:179]:
```suggestion
      it('should return the Metadata entry', async () => {
```

File: src/metadata/metadata.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+
+      const mockAsset = new MockAsset();
+      when(mockAssetsService.findOne).calledWith(ticker).mockResolvedValue(mockAsset);
+
+      mockTransactionsService.submit.mockResolvedValue({
+        result: mockMetadataEntry,
+        transactions: [mockTransaction],
+      });
+
+      const signer = 'signer';
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:209]:
```suggestion
```
import `signer` from `src/test-utils/consts.ts` instead

File: src/metadata/metadata.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        .calledWith({ ticker, type, id })
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        .mockResolvedValue(mockMetadataEntry as any);
+
+      mockTransactionsService.submit.mockResolvedValue({
+        result: mockMetadataEntry,
+        transactions: [mockTransaction],
+      });
+
+      const signer = 'signer';
```
Comment (polymath-eric) [inline src/metadata/metadata.service.spec.ts:253]:
```suggestion
```
import from test consts

File: src/metadata/metadata.service.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    } catch (err: unknown) {
+      if (isPolymeshError(err)) {
+        const { code, message } = err;
+        if (code === ErrorCode.DataUnavailable) {
+          throw new NotFoundException(message);
+        }
+      }
+
+      throw err;
+    }
```
Comment (polymath-eric) [inline src/metadata/metadata.service.ts:50]: This error logic is duplicated. This should work instead

```ts
    return await metadata.getOne({ type, id }).catch(handleSdkError);
```

File: src/metadata/metadata.service.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+  public async setValue(
+    params: MetadataParamsDto,
+    body: SetMetadataDto
+  ): ServiceReturn<MetadataEntry> {
+    const { signer, webhookUrl, ...rest } = body;
+
+    const { set } = await this.findOne(params);
+
+    return this.transactionsService.submit(set, rest as SetMetadataParams, {
```
Comment (polymath-eric) [inline src/metadata/metadata.service.ts:74]: The `as SetMetadataParams` makes me a bit nervous, but its not obvious to me how to get rid of it.

If it needs to stay than so be it.

File: src/metadata/models/metadata-entry.model.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+import { ApiProperty } from '@nestjs/swagger';
+import { BigNumber } from '@polymeshassociation/polymesh-sdk';
+import { MetadataType } from '@polymeshassociation/polymesh-sdk/types';
+
+import { FromBigNumber } from '~/common/decorators/transformation';
+
+export class MetadataEntryModel {
+  @ApiProperty({
+    description: 'Ticker of the Asset for which this is the metadata',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-entry.model.ts:11]:
```suggestion
    description: 'The ticker of the Asset for which this is the Metadata for',
```

File: src/metadata/models/metadata-entry.model.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    type: 'string',
+    example: 'TICKER',
+  })
+  readonly asset: string;
+
+  @ApiProperty({
+    description: 'Type of metadata represented by this instance',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-entry.model.ts:18]:
```suggestion
    description: 'The type of metadata represented by this instance',
```

File: src/metadata/models/metadata-spec.model.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
+/* istanbul ignore file */
+
+import { ApiPropertyOptional } from '@nestjs/swagger';
+
+export class MetadataSpecModel {
+  @ApiPropertyOptional({
+    description: 'Off-chain specs or documentation link',
+    type: 'string',
+    example: 'https://www.someexample.com',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-spec.model.ts:9]:
```suggestion
    example: 'https://example.com',
```

File: src/metadata/models/metadata-spec.model.ts [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  })
+  readonly url?: string;
+
+  @ApiPropertyOptional({
+    description: 'Description of metadata type',
+    type: 'string',
+    example: 'https://www.someexample.com',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-spec.model.ts:16]:
```suggestion
    example: 'https://example.com',
```

File: src/metadata/models/metadata-spec.model.ts [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+export class MetadataSpecModel {
+  @ApiPropertyOptional({
+    description: 'Off-chain specs or documentation link',
+    type: 'string',
+    example: 'https://www.someexample.com',
+  })
+  readonly url?: string;
+
+  @ApiPropertyOptional({
+    description: 'Description of metadata type',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-spec.model.ts:14]:
```suggestion
    description: 'Description of the Metadata type',
```

File: src/metadata/models/metadata-spec.model.ts [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+    type: 'string',
+    example: 'https://www.someexample.com',
+  })
+  readonly description?: string;
+
+  @ApiPropertyOptional({
+    description: 'SCALE encoded `AssetMetadataTypeDef`',
+    type: 'string',
+    example: 'https://www.someexample.com',
```
Comment (polymath-eric) [inline src/metadata/models/metadata-spec.model.ts:23]:
```suggestion
    example: 'https://example.com',
```

File: src/metadata/models/metadata-value.model.ts [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+  })
+  readonly lockStatus: MetadataLockStatus;
+
+  @ApiPropertyOptional({
+    description:
+      'Date till which the Metadata value will be locked. This only applies when `status` is `LockedUntil`',
+    type: 'string',
+    example: new Date('05/23/2021').toISOString(),
+  })
+  readonly lockedUntil?: Date;
```
Comment (polymath-eric) [inline src/metadata/models/metadata-value.model.ts:35]: To me this seems a bit inconsistent with `expiry` which is nullable.

I am OK with it as long as you know the difference in behavior and this is what you want: https://dev.to/jsmates/why-the-undefined-variable-is-not-sent-in-a-json-api-response-18k3#:~:text=Undefined%20is%20not%20a%20valid,valid%20JSON%20and%20vice%20versa.&text=All%20other%20values%20will%20be,undefined%20won't%20be%20sent.

File: src/transactions/transactions.service.ts [hunk 25]
```diff
@@ -87,7 +87,7 @@ export class TransactionsService {
         );
       }
     } catch (error) {
-      handleSdkError(error);
+      handleSdkError(error as Error);
```
Comment (polymath-eric) [inline src/transactions/transactions.service.ts:90]:
```suggestion
      handleSdkError(error);
```

This shouldn't be needed and doesn't provide any additional type safety

File: src/transactions/transactions.util.ts [hunk 26]
```diff
@@ -97,7 +97,7 @@ export async function processTransaction<
       transactions: [assembleTransactionResponse(procedure)],
     };
   } catch (err) /* istanbul ignore next: not worth the trouble */ {
-    handleSdkError(err);
+    handleSdkError(err as Error);
```
Comment (polymath-eric) [inline src/transactions/transactions.util.ts:100]:
```suggestion
    handleSdkError(err);
```

File: src/transactions/transactions.service.ts (hunk 25 above)
Comment (polymath-eric) [reply src/transactions/transactions.service.ts:90]: Seems like something changed that makes us need it now with new packages.

OK to leave

File: src/metadata/dto/create-metadata.dto.ts [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+  @ValidateNested()
+  @Type(() => MetadataSpecDto)
+  readonly specs: MetadataSpecDto;
+
+  @ApiPropertyOptional({
+    description: 'Value for the Asset Metadata',
+    type: 'string',
+    example: 'SOME_VALUE',
+  })
+  @ValidateIf(({ details, value }: CreateMetadataDto) => !!value || !!details)
```
[the remark this answers is not in the document]
Comment (prashantasdeveloper) [reply src/metadata/dto/create-metadata.dto.ts:32]: My idea was it to make it err if details are provided without value. `details` field actually attributes to `value` part of metadata.

File: src/metadata/dto/metadata-spec.dto.ts [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+    example: 'Some description',
+  })
+  @IsOptional()
+  @IsString()
+  readonly description?: string;
+
+  @ApiPropertyOptional({
+    description: 'SCALE encoded `AssetMetadataTypeDef`',
+    type: 'string',
+    example: 'Some example',
```
[the remark this answers is not in the document]
Comment (prashantasdeveloper) [reply src/metadata/dto/metadata-spec.dto.ts:28]: I have added the link using markdown

File: src/metadata/metadata.controller.ts (hunk 6 above)
Comment (prashantasdeveloper) [reply src/metadata/metadata.controller.ts:144]: I have changed the text, but to be clear what we can do with this API is that we can 
* only set value
* set value along with details
* only set details (this requires the value to be already set)

File: src/metadata/models/metadata-value.model.ts (hunk 24 above)
Comment (prashantasdeveloper) [reply src/metadata/models/metadata-value.model.ts:35]: Wrt to SDK, `null` signifies never in such scenarios (more of like eternity). LockedUntil is only present when status would say `LockedUntil`, hence the `?` part. `null` value here would mean that LockedUntil eternity which isn't the case. So for this scenario it fits correctly IMO

File: src/metadata/metadata.service.ts (hunk 17 above)
Comment (prashantasdeveloper) [reply src/metadata/metadata.service.ts:74]: This is required in this case. To get rid of it we just need to create a new object (of type SetMetadataParams) with all the keys & values within rest which seems redundant

File: src/metadata/metadata.controller.spec.ts (hunk 2 above)
Comment (prashantasdeveloper) [reply src/metadata/metadata.controller.spec.ts:110]: After Toms change, we can refactor couples of files to match this pattern. For now I'll leave like this

File: src/metadata/metadata.controller.spec.ts [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+    expect(controller).toBeDefined();
+  });
+
+  describe('getMetadata', () => {
+    it('should return the list of all metadata for a given ticker', async () => {
+      const mockMetadataEntry = new MockMetadataEntry();
+      when(mockService.findAll)
+        .calledWith(ticker)
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        .mockResolvedValue([mockMetadataEntry as any]);
```
[the remark this answers is not in the document]
Comment (prashantasdeveloper) [reply src/metadata/metadata.controller.spec.ts:56]: That's how we have been setting up all the test suites, rather it be SDK or REST API. `TICKER` serves as a default value for Asset.ticker and we have created the metadata for the same as well. So I don't think that should be a problem. Said that I'll change the code to use createMock instead as well

File: src/test-utils/service-mocks.ts [hunk 30]
```diff
@@ -204,6 +205,11 @@ export const mockTrustedClaimIssuersServiceProvider: ValueProvider<TrustedClaimI
   useValue: createMock<TrustedClaimIssuersService>(),
 };
 
+export const mockMetadataServiceProvider: ValueProvider<MetadataService> = {
```
Comment (VictorVicente) [inline src/test-utils/service-mocks.ts:208]:
```suggestion
export const MockMetadataServiceProvider: ValueProvider<MetadataService> = {
```

Comment (polymath-eric) [reply src/test-utils/service-mocks.ts:208]: This is a plain JavaScript object, so camel case seems right to me

Pascal case should be used for types and classes

File: src/metadata/dto/create-metadata.dto.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (polymath-eric) [reply src/metadata/dto/create-metadata.dto.ts:32]: Makes sense. I was thinking details was more related to the key part of the metadata

File: src/metadata/metadata.controller.ts (hunk 4 above)
Comment (polymath-eric) [reply src/metadata/metadata.controller.ts:122]: Fair enough

**Review by polymath-eric — APPROVED** [review]

**Pull request merged**